File: apps/medousa-home/src/lib/utils/environmentDefault.ts
import type { EnvironmentSpec, SurfaceDef } from "$lib/types/environment";
import {
  SAFETY_SURFACE_RUNTIME,
  SAFETY_SURFACE_SETTINGS,
} from "$lib/types/environment";

const DEFAULT_PROFILE_ID = "personal";
const DEFAULT_PRESET_ID = "default";

function defaultShellChrome() {
  return {
    mobile: {
      defaultHome: "home",
      askEntry: "inline" as const,
      tabBar: "full" as const,
    },
    desktop: null,
  };
}

function defaultSurfaces() {
  const builtin: Array<{
    id: string;
    label: string;
    icon: string;
    builtinId: string;
    mobileTab?: string;
  }> = [
    { id: "home", label: "Home", icon: "home", builtinId: "home", mobileTab: "home" },
    { id: "chat", label: "Chat", icon: "message-circle", builtinId: "chat", mobileTab: "chat" },
    { id: "peers", label: "Peers", icon: "users", builtinId: "peers" },
    { id: "work", label: "Work", icon: "layout-grid", builtinId: "work" },
    { id: "code", label: "Code", icon: "code-2", builtinId: "code" },
    { id: "library", label: "Workspace", icon: "notebook-text", builtinId: "library" },
    { id: "notes", label: "Notes", icon: "notebook-text", builtinId: "notes", mobileTab: "notes" },
    { id: "files", label: "Files", icon: "folder", builtinId: "files" },
    { id: "artifacts", label: "Artifacts", icon: "sparkles", builtinId: "artifacts" },
    { id: "calendar", label: "Calendar", icon: "calendar-days", builtinId: "calendar" },
    { id: "web", label: "Web", icon: "globe", builtinId: "web", mobileTab: "web" },
    { id: "map", label: "Map", icon: "compass", builtinId: "map" },
    { id: "workshop", label: "Capabilities", icon: "zap", builtinId: "workshop" },
    { id: "automations", label: "Automations", icon: "zap", builtinId: "automations" },
    { id: "messaging", label: "Messaging", icon: "radio", builtinId: "messaging" },
    {
      id: SAFETY_SURFACE_RUNTIME,
      label: "Runtime",
      icon: "activity",
      builtinId: SAFETY_SURFACE_RUNTIME,
    },
    {
      id: SAFETY_SURFACE_SETTINGS,
      label: "Settings",
      icon: "settings",
      builtinId: SAFETY_SURFACE_SETTINGS,
    },
  ];

  return builtin.map((entry) => ({
    id: entry.id,
    label: entry.label,
    icon: entry.icon,
    kind: "builtin" as const,
    builtinId: entry.builtinId,
    layout: "single" as const,
    slots: [],
    mobileTab: entry.mobileTab ?? null,
  }));
}

function codeSurfaceDef(): SurfaceDef {
  return (
    defaultSurfaces().find((surface) => surface.id === "code") ?? {
      id: "code",
      label: "Code",
      icon: "code-2",
      kind: "builtin",
      builtinId: "code",
      layout: "single",
      slots: [],
      mobileTab: null,
    }
  );
}

function placeCodeAfterWork(surfaceIds: string[]): string[] {
  if (surfaceIds.includes("code")) return surfaceIds;
  const next = [...surfaceIds];
  const workAt = next.indexOf("work");
  if (workAt >= 0) {
    next.splice(workAt + 1, 0, "code");
    return next;
  }
  const libraryAt = next.indexOf("library");
  if (libraryAt >= 0) {
    next.splice(libraryAt, 0, "code");
    return next;
  }
  next.push("code");
  return next;
}

/** Ensure Code has a first-class destination on older saved layouts. */
export function ensureCodeSurfaceInSpec(spec: EnvironmentSpec): EnvironmentSpec {
  const hasCode = spec.surfaces.some((surface) => surface.id === "code");
  let surfaces = [...spec.surfaces];
  if (!hasCode) {
    const workAt = surfaces.findIndex((surface) => surface.id === "work");
    const libraryAt = surfaces.findIndex((surface) => surface.id === "library");
    const insertAt = workAt >= 0 ? workAt + 1 : libraryAt >= 0 ? libraryAt : surfaces.length;
    surfaces.splice(insertAt, 0, codeSurfaceDef());
  }

  // Only seed preset membership while introducing the SurfaceDef. Once the
  // definition exists, absence from a preset is an intentional nav choice.
  const layoutPresets = hasCode
    ? (spec.layoutPresets ?? [])
    : (spec.layoutPresets ?? []).map((preset) => ({
        ...preset,
        surfaces: placeCodeAfterWork(preset.surfaces),
      }));
  const surfacesChanged =
    surfaces.length !== spec.surfaces.length ||
    surfaces.some((surface, index) => surface.id !== spec.surfaces[index]?.id);
  const presetsChanged = (spec.layoutPresets ?? []).some((preset, index) => {
    const next = layoutPresets[index];
    return !next || preset.surfaces.join("\0") !== next.surfaces.join("\0");
  });
  if (!surfacesChanged && !presetsChanged) return spec;
  return {
    ...spec,
    surfaces,
    layoutPresets: layoutPresets.length > 0 ? layoutPresets : spec.layoutPresets,
  };
}

const LIBRARY_SPLIT_IDS = ["notes", "files", "artifacts"] as const;

function librarySplitSurfaceDef(id: (typeof LIBRARY_SPLIT_IDS)[number]): SurfaceDef {
  const found = defaultSurfaces().find((surface) => surface.id === id);
  if (found) return found;
  const fallback: Record<(typeof LIBRARY_SPLIT_IDS)[number], SurfaceDef> = {
    notes: {
      id: "notes",
      label: "Notes",
      icon: "notebook-text",
      kind: "builtin",
      builtinId: "notes",
      layout: "single",
      slots: [],
      mobileTab: "notes",
    },
    files: {
      id: "files",
      label: "Files",
      icon: "folder",
      kind: "builtin",
      builtinId: "files",
      layout: "single",
      slots: [],
      mobileTab: null,
    },
    artifacts: {
      id: "artifacts",
      label: "Artifacts",
      icon: "sparkles",
      kind: "builtin",
      builtinId: "artifacts",
      layout: "single",
      slots: [],
      mobileTab: null,
    },
  };
  return fallback[id];
}

/** Replace a single `library` slot with notes/files/artifacts, preserving order. */
function replaceLibraryWithSplit(surfaceIds: string[]): string[] {
  if (!surfaceIds.includes("library")) return surfaceIds;
  const next: string[] = [];
  for (const id of surfaceIds) {
    if (id === "library") {
      for (const splitId of LIBRARY_SPLIT_IDS) {
        if (!next.includes(splitId)) next.push(splitId);
      }
      continue;
    }
    next.push(id);
  }
  return next;
}

/**
 * Promote Library's three explorer modes to first-class destinations on older
 * saved layouts. Keeps `library` as the internal LME tab host.
 */
export function ensureLibrarySplitInSpec(spec: EnvironmentSpec): EnvironmentSpec {
  let surfaces = [...spec.surfaces];
  for (const id of LIBRARY_SPLIT_IDS) {
    if (!surfaces.some((surface) => surface.id === id)) {
      const libraryAt = surfaces.findIndex((surface) => surface.id === "library");
      const insertAt = libraryAt >= 0 ? libraryAt + 1 : surfaces.length;
      surfaces.splice(insertAt, 0, librarySplitSurfaceDef(id));
    }
  }

  const layoutPresets = (spec.layoutPresets ?? []).map((preset) => ({
    ...preset,
    surfaces: replaceLibraryWithSplit(preset.surfaces),
  }));

  const surfacesChanged =
    surfaces.length !== spec.surfaces.length ||
    surfaces.some((surface, index) => surface.id !== spec.surfaces[index]?.id);
  const presetsChanged = (spec.layoutPresets ?? []).some((preset, index) => {
    const next = layoutPresets[index];
    if (!next) return true;
    if (preset.surfaces.length !== next.surfaces.length) return true;
    return preset.surfaces.some((id, i) => id !== next.surfaces[i]);
  });

  if (!surfacesChanged && !presetsChanged) return spec;
  return {
    ...spec,
    surfaces,
    layoutPresets: layoutPresets.length > 0 ? layoutPresets : spec.layoutPresets,
  };
}

function peersSurfaceDef(): SurfaceDef {
  return (
    defaultSurfaces().find((surface) => surface.id === "peers") ?? {
      id: "peers",
      label: "Peers",
      icon: "users",
      kind: "builtin",
      builtinId: "peers",
      layout: "single",
      slots: [],
      mobileTab: null,
    }
  );
}

/** Insert peers next to chat when missing from a preset surface list. */
function ensurePeersInPresetSurfaces(surfaceIds: string[]): string[] {
  if (surfaceIds.includes("peers")) return surfaceIds;
  const next = [...surfaceIds];
  const chatAt = next.indexOf("chat");
  if (chatAt >= 0) {
    next.splice(chatAt + 1, 0, "peers");
    return next;
  }
  const messagingIndex = next.indexOf("messaging");
  if (messagingIndex >= 0) {
    next.splice(messagingIndex, 0, "peers");
    return next;
  }
  next.push("peers");
  return next;
}

/** Ensure Peers exists on older specs. Preserves operator-chosen rail order. */
export function ensurePeersSurfaceInSpec(spec: EnvironmentSpec): EnvironmentSpec {
  const hasPeers = spec.surfaces.some((surface) => surface.id === "peers");
  let surfaces = [...spec.surfaces];

  if (!hasPeers) {
    const chatIndex = surfaces.findIndex((surface) => surface.id === "chat");
    const insertAt = chatIndex >= 0 ? chatIndex + 1 : surfaces.length;
    surfaces.splice(insertAt, 0, peersSurfaceDef());
  }

  const layoutPresets = hasPeers
    ? (spec.layoutPresets ?? [])
    : (spec.layoutPresets ?? []).map((preset) => ({
        ...preset,
        surfaces: ensurePeersInPresetSurfaces(preset.surfaces),
      }));

  const surfacesChanged =
    surfaces.length !== spec.surfaces.length ||
    surfaces.some((surface, index) => surface.id !== spec.surfaces[index]?.id);
  const presetsChanged = (spec.layoutPresets ?? []).some((preset, index) => {
    const next = layoutPresets[index];
    if (!next) return true;
    if (preset.surfaces.length !== next.surfaces.length) return true;
    return preset.surfaces.some((id, i) => id !== next.surfaces[i]);
  });

  if (!surfacesChanged && !presetsChanged) {
    return spec;
  }

  return {
    ...spec,
    surfaces,
    layoutPresets: layoutPresets.length > 0 ? layoutPresets : spec.layoutPresets,
  };
}

function calendarSurfaceDef(): SurfaceDef {
  return (
    defaultSurfaces().find((surface) => surface.id === "calendar") ?? {
      id: "calendar",
      label: "Calendar",
      icon: "calendar-days",
      kind: "builtin",
      builtinId: "calendar",
      layout: "single",
      slots: [],
      mobileTab: null,
    }
  );
}

function placeCalendarAfterLibrary(surfaceIds: string[]): string[] {
  if (surfaceIds.includes("calendar")) return surfaceIds;
  const next = [...surfaceIds];
  const libraryAt = next.indexOf("library");
  if (libraryAt >= 0) {
    next.splice(libraryAt + 1, 0, "calendar");
    return next;
  }
  const webAt = next.indexOf("web");
  if (webAt >= 0) {
    next.splice(webAt, 0, "calendar");
    return next;
  }
  next.push("calendar");
  return next;
}

/** Ensure Calendar exists after Library in the rail. */
export function ensureCalendarSurfaceInSpec(spec: EnvironmentSpec): EnvironmentSpec {
  const hasCalendar = spec.surfaces.some((surface) => surface.id === "calendar");
  let surfaces = [...spec.surfaces];

  if (!hasCalendar) {
    const libraryIndex = surfaces.findIndex((surface) => surface.id === "library");
    const insertAt = libraryIndex >= 0 ? libraryIndex + 1 : surfaces.length;
    surfaces.splice(insertAt, 0, calendarSurfaceDef());
  }

  const layoutPresets = hasCalendar
    ? (spec.layoutPresets ?? [])
    : (spec.layoutPresets ?? []).map((preset) => ({
        ...preset,
        surfaces: placeCalendarAfterLibrary(preset.surfaces),
      }));

  const surfacesChanged =
    surfaces.length !== spec.surfaces.length ||
    surfaces.some((surface, index) => surface.id !== spec.surfaces[index]?.id);
  const presetsChanged = (spec.layoutPresets ?? []).some((preset, index) => {
    const next = layoutPresets[index];
    if (!next) return true;
    if (preset.surfaces.length !== next.surfaces.length) return true;
    return preset.surfaces.some((id, i) => id !== next.surfaces[i]);
  });

  if (!surfacesChanged && !presetsChanged) {
    return spec;
  }

  return {
    ...spec,
    surfaces,
    layoutPresets: layoutPresets.length > 0 ? layoutPresets : spec.layoutPresets,
  };
}

function mapSurfaceDef(): SurfaceDef {
  return (
    defaultSurfaces().find((surface) => surface.id === "map") ?? {
      id: "map",
      label: "Map",
      icon: "compass",
      kind: "builtin",
      builtinId: "map",
      layout: "single",
      slots: [],
      mobileTab: null,
    }
  );
}

function replaceRetiredContextSurface(
  surfaceIds: string[],
  seedMissingMap: boolean,
): string[] {
  const hadContext = surfaceIds.includes("context");
  const withoutContext = surfaceIds.filter((id) => id !== "context");
  if (withoutContext.includes("map")) return withoutContext;
  if (!hadContext && !seedMissingMap) return withoutContext;
  const next = [...withoutContext];
  const libraryAt = next.indexOf("library");
  if (libraryAt >= 0) {
    next.splice(libraryAt + 1, 0, "map");
    return next;
  }
  const webAt = next.indexOf("web");
  if (webAt >= 0) {
    next.splice(webAt + 1, 0, "map");
    return next;
  }
  next.push("map");
  return next;
}

/** Ensure Map exists; strip retired Context surface from specs/presets. */
export function ensureMapSurfaceInSpec(spec: EnvironmentSpec): EnvironmentSpec {
  let surfaces = spec.surfaces.filter((surface) => surface.id !== "context");
  const hasMap = surfaces.some((surface) => surface.id === "map");

  if (!hasMap) {
    const libraryIndex = surfaces.findIndex((surface) => surface.id === "library");
    const insertAt = libraryIndex >= 0 ? libraryIndex + 1 : surfaces.length;
    surfaces = [...surfaces];
    surfaces.splice(insertAt, 0, mapSurfaceDef());
  }

  const layoutPresets = (spec.layoutPresets ?? []).map((preset) => ({
    ...preset,
    surfaces: replaceRetiredContextSurface(preset.surfaces, !hasMap),
  }));

  const surfacesChanged =
    surfaces.length !== spec.surfaces.length ||
    surfaces.some((surface, index) => surface.id !== spec.surfaces[index]?.id);
  const presetsChanged = (spec.layoutPresets ?? []).some((preset, index) => {
    const next = layoutPresets[index];
    if (!next) return true;
    if (preset.surfaces.length !== next.surfaces.length) return true;
    return preset.surfaces.some((id, i) => id !== next.surfaces[i]);
  });

  if (!surfacesChanged && !presetsChanged) {
    return spec;
  }

  return {
    ...spec,
    surfaces,
    layoutPresets: layoutPresets.length > 0 ? layoutPresets : spec.layoutPresets,
  };
}

export function defaultEnvironmentSpec(
  profileId = DEFAULT_PROFILE_ID,
): EnvironmentSpec {
  const now = new Date().toISOString();
  const surfaces = defaultSurfaces();
  return {
    version: 1,
    profileId,
    surfaces,
    components: [],
    layoutPresets: [
      {
        id: DEFAULT_PRESET_ID,
        label: "Default",
        active: true,
        // `library` stays defined as the LME host but is not a rail door.
        surfaces: surfaces
          .map((surface) => surface.id)
          .filter((id) => id !== "library"),
        shellChrome: defaultShellChrome(),
      },
      {
        id: "focus",
        label: "Focus",
        active: false,
        surfaces: [
          "chat",
          "peers",
          "work",
          "code",
          "notes",
          "web",
          "files",
          "artifacts",
          "map",
          SAFETY_SURFACE_SETTINGS,
          SAFETY_SURFACE_RUNTIME,
        ],
        shellChrome: defaultShellChrome(),
      },
    ],
    activePresetId: DEFAULT_PRESET_ID,
    shellChrome: defaultShellChrome(),
    theme: null,
    updatedAt: now,
    updatedBy: "system",
  };
}

File: crates/medousa-types/src/environment_default.rs
//! Default environment spec matching the current hardcoded Home experience.

use chrono::Utc;

use crate::environment::{
    ComponentDef, ComponentType, EnvironmentSpec, LayoutPreset, MobileAskEntry, MobileTabBar,
    ShellChromeDef, ShellChromeMobile, SurfaceDef, SurfaceKind, SurfaceLayout, UiPresentation,
    CHROME_ACTION_OPEN_ASK, ENVIRONMENT_SPEC_VERSION, SAFETY_SURFACE_RUNTIME,
    SAFETY_SURFACE_SETTINGS,
};

pub const DEFAULT_PROFILE_ID: &str = "personal";
pub const DEFAULT_PRESET_ID: &str = "default";
pub const FOCUS_PRESET_ID: &str = "focus";

pub fn default_environment_spec(profile_id: impl Into<String>) -> EnvironmentSpec {
    let profile_id = profile_id.into();
    let now = Utc::now();
    EnvironmentSpec {
        version: ENVIRONMENT_SPEC_VERSION,
        profile_id: profile_id.clone(),
        surfaces: default_surfaces(),
        components: Vec::new(),
        layout_presets: Some(vec![
            LayoutPreset {
                id: DEFAULT_PRESET_ID.to_string(),
                label: "Default".to_string(),
                active: true,
                surfaces: default_surface_ids(),
                shell_chrome: Some(default_shell_chrome()),
            },
            LayoutPreset {
                id: FOCUS_PRESET_ID.to_string(),
                label: "Focus".to_string(),
                active: false,
                surfaces: vec![
                    "chat".to_string(),
                    "peers".to_string(),
                    "work".to_string(),
                    "library".to_string(),
                    SAFETY_SURFACE_SETTINGS.to_string(),
                    SAFETY_SURFACE_RUNTIME.to_string(),
                ],
                shell_chrome: Some(default_shell_chrome()),
            },
        ]),
        active_preset_id: Some(DEFAULT_PRESET_ID.to_string()),
        shell_chrome: Some(default_shell_chrome()),
        theme: None,
        updated_at: now,
        updated_by: "system".to_string(),
    }
}

pub fn default_shell_chrome() -> ShellChromeDef {
    ShellChromeDef {
        mobile: Some(ShellChromeMobile {
            default_home: Some("home".to_string()),
            ask_entry: Some(MobileAskEntry::Inline),
            tab_bar: Some(MobileTabBar::Full),
        }),
        desktop: None,
    }
}

pub fn default_surface_ids() -> Vec<String> {
    vec![
        "home".to_string(),
        "chat".to_string(),
        "peers".to_string(),
        "work".to_string(),
        "library".to_string(),
        "calendar".to_string(),
        "web".to_string(),
        "context".to_string(),
        "workshop".to_string(),
        "automations".to_string(),
        "messaging".to_string(),
        SAFETY_SURFACE_RUNTIME.to_string(),
        SAFETY_SURFACE_SETTINGS.to_string(),
    ]
}

pub fn default_surfaces() -> Vec<SurfaceDef> {
    let builtin = [
        ("home", "Home", "home", Some("home"), Some("home")),
        ("chat", "Chat", "message-circle", Some("chat"), Some("chat")),
        ("peers", "Peers", "users", Some("peers"), None),
        ("work", "Work", "layout-grid", Some("work"), None),
        ("library", "Library", "book-open", Some("library"), Some("notes")),
        ("calendar", "Calendar", "calendar-days", Some("calendar"), None),
        ("web", "Web", "globe", Some("web"), Some("web")),
        ("context", "Context", "orbit", Some("context"), None),
        ("workshop", "Capabilities", "zap", Some("workshop"), None),
        ("automations", "Automations", "calendar", Some("automations"), None),
        ("messaging", "Messaging", "radio", Some("messaging"), None),
        (
            SAFETY_SURFACE_RUNTIME,
            "Runtime",
            "activity",
            Some(SAFETY_SURFACE_RUNTIME),
            None,
        ),
        (
            SAFETY_SURFACE_SETTINGS,
            "Settings",
            "settings",
            Some(SAFETY_SURFACE_SETTINGS),
            None,
        ),
    ];

    builtin
        .into_iter()
        .map(|(id, label, icon, builtin_id, mobile_tab)| SurfaceDef {
            id: id.to_string(),
            label: label.to_string(),
            icon: icon.to_string(),
            kind: SurfaceKind::Builtin,
            builtin_id: builtin_id.map(str::to_string),
            layout: SurfaceLayout::Single,
            slots: vec![],
            mobile_tab: mobile_tab.map(str::to_string),
            layout_root: None,
        })
        .collect()
}

/// Phase 1 proof demo — custom home, FAB ask, presentation + chrome_action components.
pub fn writing_studio_demo_spec(profile_id: impl Into<String>) -> EnvironmentSpec {
    let mut spec = default_environment_spec(profile_id);
    spec.surfaces.push(SurfaceDef {
        id: "writing-studio".to_string(),
        label: "Writing studio".to_string(),
        icon: "pen-line".to_string(),
        kind: SurfaceKind::Custom,
        builtin_id: None,
        layout: SurfaceLayout::Dashboard,
        slots: vec![],
        mobile_tab: None,
        layout_root: None,
    });

    if let Some(presets) = &mut spec.layout_presets {
        for preset in presets.iter_mut() {
            if preset.id == DEFAULT_PRESET_ID && !preset.surfaces.iter().any(|id| id == "writing-studio") {
                preset.surfaces.push("writing-studio".to_string());
            }
        }
    }

    spec.shell_chrome = Some(ShellChromeDef {
        mobile: Some(ShellChromeMobile {
            default_home: Some("writing-studio".to_string()),
            ask_entry: Some(MobileAskEntry::Fab),
            tab_bar: Some(MobileTabBar::Full),
        }),
        desktop: None,
    });

    spec.components.push(ComponentDef {
        id: "writing-manuscript".to_string(),
        component_type: ComponentType::Presentation,
        surface_id: "writing-studio".to_string(),
        slot: "main".to_string(),
        label: Some("Manuscript".to_string()),
        config: serde_json::json!({ "artifactId": "art-writing-demo" }),
        presentation: Some(UiPresentation::Inline),
        feeds: vec![],
        updated_at: Some(spec.updated_at),
    });
    spec.components.push(ComponentDef {
        id: "writing-ask-fab".to_string(),
        component_type: ComponentType::ChromeAction,
        surface_id: "writing-studio".to_string(),
        slot: "fab".to_string(),
        label: Some("Ask".to_string()),
        config: serde_json::json!({ "action": CHROME_ACTION_OPEN_ASK }),
        presentation: None,
        feeds: vec![],
        updated_at: Some(spec.updated_at),
    });
    spec.updated_by = "demo".to_string();
    spec
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::environment_validate::is_valid_environment_spec;

    #[test]
    fn default_spec_includes_safety_surfaces() {
        let spec = default_environment_spec(DEFAULT_PROFILE_ID);
        let ids: Vec<_> = spec.surfaces.iter().map(|s| s.id.as_str()).collect();
        assert!(ids.contains(&SAFETY_SURFACE_SETTINGS));
        assert!(ids.contains(&SAFETY_SURFACE_RUNTIME));
    }

    #[test]
    fn writing_studio_demo_spec_validates() {
        let spec = writing_studio_demo_spec(DEFAULT_PROFILE_ID);
        assert!(is_valid_environment_spec(&spec));
        assert_eq!(
            spec.shell_chrome
                .as_ref()
                .and_then(|c| c.mobile.as_ref())
                .and_then(|m| m.default_home.as_deref()),
            Some("writing-studio")
        );
    }
}
